### PR TITLE
add wifi connection and azure function

### PR DESCRIPTION
**Pull request by ChenTanyi**

### CODE REVIEW

File: app.ino [hunk 1]
```diff
@@ -97,4 +148,15 @@ void checkMagnetometerStatus()
             }
         }
     }
-}
\ No newline at end of file
+}
+
+/*void generateMessage(char* message)
```
Comment (formulahendry) [inline app.ino]: This could be removed.

File: app.ino [hunk 2]
```diff
@@ -1,4 +1,7 @@
 #include "LIS2MDLSensor.h"
+#include "AZ3166WiFi.h"
+#include "iothub_client_sample_mqtt.h"
+#include <ArduinoJson.h>
```
Comment (formulahendry) [inline app.ino]: I think we do not need ArduinoJson.h

File: azureFunction/function.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      "connection": "eventHubConnectionString"
+    },
+    {
+      "name": "$return",
+      "type": "sendGrid",
+      "direction": "out",
+      "apiKey": "mySendGridKey",
+      "to": "/* the address you want to send email to */",
+      "from": "/* your grid email address */",
+      "subject": "Azure news"
```
Comment (formulahendry) [inline azureFunction/function.json]: Better to change title like 'Door Status Changed'

File: iothub_client_sample_mqtt.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (formulahendry) [inline iothub_client_sample_mqtt.cpp:1]: keep this file as small as possible, remove unused code

File: azureFunction/function.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+      "name": "eventHubMessages",
+      "direction": "in",
+      "path": "%eventHubConnectionPath%",
+      "connection": "eventHubConnectionString"
+    },
+    {
+      "name": "$return",
+      "type": "sendGrid",
+      "direction": "out",
+      "apiKey": "mySendGridKey",
```
Comment (formulahendry) [inline azureFunction/function.json:14]: For those field that users need to provide, make the format like `"apiKey": "<mySendGridKey>"`, `"to": "<abcde>"`

File: app.ino [hunk 6]
```diff
@@ -50,18 +72,47 @@ void loop()
...
+            message = "Door closed";
+            curOpened = false;
         }
         else
         {
             Screen.print(0, "Door opened");
+            message = "Door opened";
+            curOpened = true;
+        }
+        //send message when status change
```
Comment (formulahendry) [inline app.ino]: Blank after `//`

File: app.ino [hunk 7]
```diff
@@ -13,24 +16,43 @@ int base_z;
...
+    if (!hasWifi)
+    {
+        delay(1000);
+        return;
+    }
+
     // read id
     lis2mdl->readId(&id);
-    Serial.printf("Id: %d\n", id);
+    //Serial.printf("Id: %d\n", id);
```
Comment (formulahendry) [inline app.ino]: Delete if not used

File: app.ino [hunk 8]
```diff
@@ -13,24 +16,43 @@ int base_z;
...
+
     // getMAxes
     lis2mdl->getMAxes(axes);
-    Serial.printf("Axes: x - %d, y - %d, z - %d\n", axes[0], axes[1], axes[2]);
+    //Serial.printf("Axes: x - %d, y - %d, z - %d\n", axes[0], axes[1], axes[2]);
```
Comment (formulahendry) [inline app.ino]: I think we can keep it.

**Review by formulahendry — CHANGES_REQUESTED** [review]

File: azureFunction/function.json (hunk 5 above)
Comment (ChenTanyi) [reply azureFunction/function.json:14]: The apiKey should be the name of api key stored in app settings like connection, so I just keep it.

File: iothub_client_sample_mqtt.cpp [hunk 9]
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "AzureIotHub.h"
+#include "Arduino.h"
+#include "EEPROMInterface.h"
+#include "iothub_client_sample_mqtt.h"
+
+static IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle;
+static bool messagePending = false;
```
Comment (formulahendry) [inline iothub_client_sample_mqtt.cpp]: In this scenario, 'messagePending' is not useful, we could remove it

**Review by formulahendry — CHANGES_REQUESTED** [review]

**Review by formulahendry — APPROVED** [review]

**Pull request merged**